### PR TITLE
Users of any kind can brows reported issues

### PR DESCRIPTION
**Pull request by mis6ko13**
Issue - https://github.com/Kotany-IF070-WebUI-RoR/NewVasjuky/issues/35
Issue - https://github.com/Kotany-IF070-WebUI-RoR/NewVasjuky/issues/68
- [ ]  @igorkasyanchuk 
- [ ]  @krisberry
- [x]  @simplews 

Fixed issues-feed layout at home page.
Implemented site.com/issues page with pagination + added localization.
Added front-end validation of attachment uploads, implemented picture resizing.

### CODE REVIEW

File: config/environments/development.rb [hunk 1]
```diff
@@ -41,7 +41,7 @@
   # Debug mode disables concatenation and preprocessing of assets.
   # This option may cause significant delays in view rendering with a large
   # number of complex assets.
-  config.assets.debug = true
+  config.assets.debug = false
```
Comment (Kmalynss) [inline config/environments/development.rb]: Why did you put ```false``` here?

Comment (mis6ko13) [inline config/environments/development.rb]: I did that for faster views rendering.

File: app/assets/javascripts/flash.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+UnobtrusiveFlash.flashOptions['timeout'] = 2000; // milliseconds
+$('.unobtrusive-flash-container').click(function(){
```
Comment (igorkasyanchuk) [inline app/assets/javascripts/flash.js]: wrap this code in $().ready because $ may not be loaded

File: app/assets/stylesheets/home.scss [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    color: whitesmoke;
+  }
+}
+
+.center:hover {
+  background-color: dodgerblue;
+  border-color: dodgerblue;
+}
+
+.nav-pills, .feed {
```
Comment (igorkasyanchuk) [inline app/assets/stylesheets/home.scss]: too global change. you are overriding all nav-pills in whole app
add ID or class to your element and specify CSS classes exactly to it

File: app/views/home/index.html.slim [hunk 4]
```diff
@@ -1 +1,23 @@
...
+        = link_to 'Повідомити про проблему', new_issue_path
+  ul.nav.nav-pills role="tablist"
+    li role="presentation"
+        span.badge #{@count}
+        span.phrase Проблем додано
+    li role="presentation"
+        span.badge #{@approved_count}
+        span.phrase Проблем вирішено
+  .row
+    h3.indent Останні додані:
```
Comment (igorkasyanchuk) [inline app/views/home/index.html.slim]: remove this class "indent" completely

inside ROW you MUST put content inside COL element. and you will not need this class.

this is the rule!

File: app/assets/stylesheets/issues.scss [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+.issues-section {
+  margin-top: 40px;
+}
+.issue-title {
+  margin-top: 10px;
+}
+
+.pagination {
```
Comment (igorkasyanchuk) [inline app/assets/stylesheets/issues.scss]: too global change. use specific class or id of parent for this

File: app/assets/stylesheets/main.scss [hunk 6]
```diff
@@ -1373,13 +1373,11 @@ fieldset[disabled] {
   border: none;
 }
 
-.notice {
-    color: darkgreen;
-}
-.alert {
-    color: red;
+.alert{
```
Comment (igorkasyanchuk) [inline app/assets/stylesheets/main.scss]: what about "notice" class? or other notifications? do you have any?

File: app/controllers/account/admin/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# Encoding: utf-8
+# frozen_string_literal: true
+module Account
+  module Admin
+    class UsersController < ApplicationController
+      def index
+        @users = User.page(params[:page]).per(20)
+        authorize @users
```
Comment (igorkasyanchuk) [inline app/controllers/account/admin/users_controller.rb]: remove authorize and remove pundit gem completely

File: app/controllers/home_controller.rb [hunk 8]
```diff
@@ -2,5 +2,10 @@
 class HomeController < ApplicationController
   skip_before_action :authenticate_user!, only: [:index]
 
-  def index; end
+  def index
+    @issues = Issue.where(approved: true)
+    @issues_feed = @issues.last(4).reverse
```
Comment (igorkasyanchuk) [inline app/controllers/home_controller.rb]: use "order" and remove reverse

File: app/views/home/index.html.slim [hunk 9]
```diff
@@ -1 +1,23 @@
...
+    - @issues_feed.each do |issue|
+      .col-md-3
+        .thumbnail= link_to issues_url
```
Comment (igorkasyanchuk) [inline app/views/home/index.html.slim]: wrong html.
you cannot put "p" inside link

ALSO what is this

`.img style="background-image:url(#{issue.attachment})"`

no inline styles.  Use image_tag

File: app/views/issues/_issue.html.slim [hunk 10]
```diff
@@ -0,0 +1,14 @@
+.media
+  .media-left.media-top
+    = link_to 'issue_path'
```
Comment (igorkasyanchuk) [inline app/views/issues/_issue.html.slim]: what is this `= link_to 'issue_path'` ??

remove inline styles below

File: app/views/issues/_issue.html.slim [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+      .issue-img.media-object style="background-image:url(#{issue.attachment})"
+  .media-body
+    = link_to 'issue_path'
+      h4.media-heading.issue-title #{issue.name}
+    span #{'Категорія: '}
+    = link_to issue.category.name, '#'
+    p.issue-desc #{issue.description}
+    .timestamp
+      span.author Автор: #{issue.user.first_name} #{issue.user.last_name}
```
Comment (igorkasyanchuk) [inline app/views/issues/_issue.html.slim]: issue.user.full_name

File: app/controllers/account/admin/users_controller.rb (hunk 7 above)
Comment (mis6ko13) [inline app/controllers/account/admin/users_controller.rb]: This one is already in progress - https://github.com/Kotany-IF070-WebUI-RoR/NewVasjuky/issues/46

File: app/assets/stylesheets/main.scss (hunk 6 above)
Comment (simplews) [inline app/assets/stylesheets/main.scss]: @igorkasyanchuk, this classes already defined in bootstrap css, so we didn't need it here. 
Also, it makes all flash text red (even if we have success or notice messages, it is not OK

File: app/views/account/admin/users/_user.html.slim [hunk 12]
```diff
@@ -8,9 +8,9 @@ tr
...
         = link_to 'Заблокувати', toggle_ban_account_admin_user_path(user), method: :patch
   td
-    - if admin? && !user.admin?
+    - if admin_change_role?(user)
       = simple_form_for(user, url: change_role_account_admin_user_path(user)) do |f|
         = f.select(:role, User.roles.keys)
         button.btn.btn-default.btn-xs = f.submit 'Змінити роль'
     - else
-      = user.role
+      = user.role
```
Comment (simplews) [inline app/views/account/admin/users/_user.html.slim]: Note that it is a good style to always put the newline as a last character

File: app/views/home/index.html.slim [hunk 13]
```diff
@@ -1 +1,24 @@
...
+      span.badge #{@approved_count}
+      span.phrase Проблем вирішено
+.row
+  .col-md-12
+    h3 Останні додані:
+  - @issues_feed.each do |issue|
+    .col-md-3.col-sm-3
+      .thumbnail.feed-issue= link_to issue
+        .attachment= image_tag issue.attachment
+        span.caption #{issue.title}
```
Comment (simplews) [inline app/views/home/index.html.slim]: same as in pp/views/account/admin/users/_user.html.slim  here

File: app/views/issues/_issue.html.slim [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+  .media-body
+    = link_to issue
+      h4.media-heading.issue-title #{issue.name}
+    span #{'Категорія: '}
+    = link_to issue.category.name, '#'
+    p.issue-about #{issue.description}
+    .timestamp
+      span.author Автор: #{issue.user.full_name}
+      br
+      span Добавлено:    #{time_ago_in_words(issue.created_at)} тому
```
Comment (simplews) [inline app/views/issues/_issue.html.slim]: Please change "Добавлено" to "Додано" and add newline

File: app/views/issues/new.html.slim [hunk 15]
```diff
@@ -17,9 +17,19 @@ h1 Надіслати нове звернення
...
         = f.input :title, label: 'Короткий опис проблеми'
-        = f.association :category, collection: Category.order(:name).all, label: 'Категорія'
+        = f.association :category, collection: Category.order(:name).all,
+                    label: 'Категорія'
 	    = f.input :location, label: 'Адреса проблеми',
                     placeholder: 'Івано-Франківськ, Сахарова, 33'
         = f.input :description, as: :text, label: 'Текст звернення'
-        = f.input :attachment, as: :file, label: 'Додаток'
+        = f.input :attachment, as: :file, label: 'Додаток',
+                    accept: 'image/jpeg,image/gif,image/png'
```
Comment (simplews) [inline app/views/issues/new.html.slim]: There will be good to ask user to upload files only in accepted file formats

File: app/views/issues/new.html.slim [hunk 16]
```diff
@@ -17,9 +17,19 @@ h1 Надіслати нове звернення
...
         = f.button :submit, class: 'btn btn-primary'
+javascript:
+  $('#issue_attachment').bind('change', function() {
+    var size_in_megabytes = this.files[0].size/1024/1024;
+    if (size_in_megabytes > 10) {
+          alert('Розмір фото повиннен бути менше 5Mб. Будь-ласка виберіть фото меншого розміру.');
```
Comment (mrkava) [inline app/views/issues/new.html.slim]: size_in_megabytes > 10 but "Розмір фото повиннен бути менше 5Mб" ???

File: db/seeds.rb [hunk 17]
```diff
@@ -159,138 +158,135 @@
...
-                             category_id: categories.index(issue[:category]),
-                             description: issue[:description],
-                             attachment: issue[:attachment],
-                             created_at: Time.now - rand(10).days)
+  FactoryGirl.create(:issue, title: issue[:title],
+                     address: issue[:address],
+                     phone: Faker::PhoneNumber.cell_phone,
+                     category: issue[:category],
+                     description: issue[:description],
+                     attachment: issue[:attachment],
```
Comment (mrkava) [inline db/seeds.rb]: Why you removed location, name, latitude, longithde? Those fields also needed

File: app/views/issues/new.html.slim (hunk 15 above)
Comment (andrjuha01) [inline app/views/issues/new.html.slim]: @simplews We have carrierwave gem, where we give the whitelist of allowed extensions.
We don't need this "accept")

**Comment by f1nwe** [comment]
Do not forget to squash commits before PR

File: db/seeds.rb (hunk 17 above)
Comment (mis6ko13) [inline db/seeds.rb]: I didn't. 
We didn't implement full issue modeling in seeds until you've added Geo-location.
Fixed!

**Comment by mis6ko13** [comment]
@igorkasyanchuk Fixed all mentioned mistakes.
Please review!

**Comment by Kmalynss** [comment]
@f1nwe as we were talking some time before, we don't have to squash commits before creating PR. Commits can be squashed with help of github at moment of merge.

File: app/views/account/admin/users/_user.html.slim [hunk 18]
```diff
@@ -2,15 +2,15 @@ tr
   td= user.email
   td= user.full_name
   td
-    - if policy(user).toggle_ban?
+    - if admin_or_moderator? && user.reporter?
       - if user.banned?
         = link_to 'Розблокувати', toggle_ban_account_admin_user_path(user), method: :patch
       - else
         = link_to 'Заблокувати', toggle_ban_account_admin_user_path(user), method: :patch
```
Comment (Kmalynss) [inline app/views/account/admin/users/_user.html.slim]: Avoid using conditions in views. Helper is a good place for such things.

**Comment by mis6ko13** [comment]
@Kmalynss fixed conditions in views.
Please review!

File: app/views/account/admin/users/_user.html.slim [hunk 19]
```diff
@@ -2,15 +2,9 @@ tr
   td= user.email
   td= user.full_name
   td
-    - if admin_or_moderator? && user.reporter?
-      - if user.banned?
-        = link_to 'Розблокувати', toggle_ban_account_admin_user_path(user), method: :patch
-      - else
-        = link_to 'Заблокувати', toggle_ban_account_admin_user_path(user), method: :patch
+    = link_to bann_user?(user), bann_path(user), method: :patch
```
Comment (krisberry) [inline app/views/account/admin/users/_user.html.slim]: What will be rendered on page if user is an admin? 
We shouldn't render this element if helpers return nil or empty string. So you can write a helper which will return whole element when it needed.

File: app/views/issues/index.html.slim [hunk 20]
```diff
@@ -0,0 +1,12 @@
+h1 Звернення
+.dropdown.issue-sort-by
+  a.dropdown-toggle href='#' data-toggle="dropdown"
+    span Сортувати
+    span.caret
+    ul.dropdown-menu role='menu'
+      li= link_to 'Спочатку нові', '#'
```
Comment (igorkasyanchuk) [inline app/views/issues/index.html.slim]: just idea for future, you can use later gem ransack for sorting

**Comment by Kmalynss** [comment]
Files changed: 198.
Looks like you changed permissions for whole project directory. Not good. Please change them back, so only rly changed files will appear in the PR.

**Comment by mis6ko13** [comment]
@Kmalynss,
fixed issue with project's directory permissions and 198 changed files.
@krisberry,
fixed codeclimate issue with "= render @issues"

**Pull request merged**